core/src/apps/webauthn/__init__.py

```diff
@@ -314,141 +423,541 @@ def to_msg(self) -> Msg:
+class ConfirmInfo:
+    def __init__(self) -> None:
+        self.app_icon = None  # type: Optional[bytes]
 
-        self.confirmed = None
-        self.workflow = self.confirm_workflow()
-        loop.schedule(self.workflow)
-        return True
+    def get_header(self) -> Optional[str]:
+        return None
+
+    def app_name(self) -> Optional[str]:
+        return None
+
+    def account_name(self) -> Optional[str]:
+        return None
 
-    def keepalive(self):
-        self.deadline = utime.ticks_ms() + _CONFIRM_TIMEOUT_MS
+    def load_icon(self, rp_id_hash: bytes) -> None:
+        from trezor import res
+        from apps.webauthn import knownapps
 
-    async def confirm_workflow(self) -> None:
         try:
-            workflow.onstart(self.workflow)
-            await self.confirm_layout()
-        finally:
-            workflow.onclose(self.workflow)
-            self.workflow = None
+            namepart = knownapps.knownapps[rp_id_hash].lower().replace(" ", "_")
+            icon = res.load("apps/webauthn/res/icon_%s.toif" % namepart)
+        except Exception as e:
+            icon = res.load("apps/webauthn/res/icon_webauthn.toif")
+            if __debug__:
+                log.exception(__name__, e)
+        self.app_icon = icon
+
+
+class ConfirmContent(ui.Control):
+    def __init__(self, info: ConfirmInfo) -> None:
+        self.info = info
+        self.repaint = True
+
+    def on_render(self) -> None:
+        if self.repaint:
+            header = self.info.get_header()
+
+            if header is None or self.info.app_icon is None:
+                return
+            ui.header(header, ui.ICON_DEFAULT, ui.GREEN, ui.BG, ui.GREEN)
+            ui.display.image((ui.WIDTH - 64) // 2, 48, self.info.app_icon)
+
+            app_name = self.info.app_name()
+            account_name = self.info.account_name()
+
+            # Dummy requests usually have some text as both app_name and account_name,
+            # in that case show the text only once.
+            if (
+                app_name is not None
+                and account_name is not None
+                and app_name != account_name
+            ):
+                text_center_trim_left(ui.WIDTH // 2, 140, app_name)
+                text_center_trim_right(ui.WIDTH // 2, 172, account_name)
+            elif account_name is not None:
+                text_center_trim_right(ui.WIDTH // 2, 156, account_name)
+            elif app_name is not None:
+                text_center_trim_left(ui.WIDTH // 2, 156, app_name)
 
-    async def confirm_layout(self) -> None:
-        from trezor.ui.confirm import Confirm, CONFIRMED
-        from trezor.ui.text import Text
+            self.repaint = False
+
+
+class State:
+    def __init__(self, cid: int, iface: io.HID) -> None:
+        self.cid = cid
+        self.iface = iface
+
+    def keepalive_status(self) -> Optional[int]:
+        return None
+
+    def timeout_ms(self) -> int:
+        return _U2F_CONFIRM_TIMEOUT_MS
```

Comment:
I think _U2F_CONFIRM_TIMEOUT_MS should be returned from U2fState instead of State.

Reply:
Decided to raise NotImplemented instead.